File: address-geocode-international-python/REST/reverse_search_rest.py
'''
Service Objects, Inc.

This module provides the reverse_search function to perform reverse geocoding
using Service Objects' AGI ReverseSearch API. It handles live/trial endpoints,
fallback logic, and JSON parsing.

Functions:
    reverse_search(latitude: float,
                longitude: float,
                search_radius: int,
                country: str,
                max_results: int,
                search_type: str,
                license_key: str,
                is_live: bool) -> dict
'''

import requests  # HTTP client for RESTful API calls

# Endpoint URLs for AV3 service
PRIMARY_URL = "https://sws.serviceobjects.com/AGI/api.svc/json/ReverseSearch"
BACKUP_URL = "https://swsbackup.serviceobjects.com/AGI/api.svc/ReverseSearch"
TRIAL_URL = "https://trial.serviceobjects.com/AGI/api.svc/json/ReverseSearch"

def reverse_search(latitude: float,
                longitude: float,
                search_radius: int,
                country: str,
                max_results: int,
                search_type: str,
                license_key: str,
                is_live: bool) -> dict:
    """
    Call AGI ReverseSearch API and return location results.

    Parameters:
        latitude (float): Latitude coordinate.
        longitude (float): Longitude coordinate.
        search_radius (int): The maximum search radius distance set in kilometers. Defaults to 1. Maximum value of 50.
        country (str): The preferred name of the country or the ISO 3166-1 Alpha-2 Country Code or the Alpha-3 Country Code.
        max_results (int): Maximum number of returned locations.
        search_type (str): The name of the type of search you want to perform for the given address or place.
        license_key (str): Service Objects license key.
        is_live (bool): True for production, False for trial endpoint.

    Returns:
        dict: Parsed JSON response with location data or error info.

    Raises:
        RuntimeError: If the API returns an error payload.
        requests.RequestException: On network/HTTP failures (trial mode).
    """

    # Prepare query parameters for AGI API
    params = {
        "Latitude": latitude,
        "Longitude": longitude,
        "SearchRadius": search_radius,
        "Country": country,
        "MaxResults": max_results,
        "SearchType": search_type,
        "LicenseKey": license_key
    }

    # Select the base URL: production vs trial
    url = PRIMARY_URL if is_live else TRIAL_URL

    try:
        # Attempt primary (or trial) endpoint first
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = getattr(response, 'Error', None)
        if not (error is None or getattr(error, 'TypeCode', None) != "3"):
            if is_live:
                # Try backup URL when live
                response = requests.get(BACKUP_URL, params=params, timeout=10)
                data = response.json()

                # If still error, propagate exception
                if "Error" in data:
                    raise RuntimeError(f"AGI ReverseSearch error: {data['Error']}")
            else:
                # Trial mode should not fallback; error is terminal
                raise RuntimeError(f"AGI trial error: {data['Error']}")
        
        # Success: return parsed JSON data
        return data

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL on network failure
                response = requests.get(BACKUP_URL, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                if "Error" in data:
                    # Both primary and backup failed; escalate
                    raise RuntimeError(f"AGI ReverseSearch backup error: {data['Error']}")
                return data
            except Exception as backup_exc:
                raise RuntimeError("AGI ReverseSearch service unreachable on both endpoints") from backup_exc

        else:
            # In trial mode, propagate the network exception
            raise RuntimeError(f"AGI trial error: {str(req_exc)}") from req_exc

File: address-geocode-international-nodejs/REST/reverse_search_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { RSResponse } from './agi_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects Address Geocode International (AGI) API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/agi/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects Address Geocode International (AGI) API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/agi/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects Address Geocode International (AGI) API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/agi/api.svc/json/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.TypeCode is not equal to '3'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.TypeCode !== '3';

/**
 * <summary>
 * Constructs a full URL for the ReverseSearch API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}ReverseSearch?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<RSResponse>">A promise that resolves to an RSResponse object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new RSResponse(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects Address Geocode International (AGI) API's ReverseSearch endpoint,
 * retrieving address or place information for given coordinates with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const ReverseSearchClient = {
    /**
     * <summary>
     * Asynchronously invokes the ReverseSearch API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.TypeCode == '3') in live mode.
     * </summary>
     * @param {string} Latitude - The latitude of the location (decimal, up to 7 decimal places). Required.
     * @param {string} Longitude - The longitude of the location (decimal, up to 7 decimal places). Required.
     * @param {string} SearchRadius - The maximum search radius in kilometers (1-50). Defaults to '1'.
     * @param {string} Country - The country name or ISO 3166-1 Alpha-2/Alpha-3 code. Optional.
     * @param {string} MaxResults - Maximum number of results (1-10). Defaults to '10'.
     * @param {string} SearchType - Type of search (e.g., 'BestMatch', 'All'). Required.
     * @param {string} LicenseKey - Your license key to use the service. Required.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers. Defaults to true.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service. Defaults to 15.
     * @returns {Promise<RSResponse>} - A promise that resolves to an RSResponse object.
     */
    async invokeAsync(Latitude, Longitude, SearchRadius = '1', Country, MaxResults = '10', SearchType, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            Latitude,
            Longitude,
            SearchRadius,
            Country,
            MaxResults,
            SearchType,
            LicenseKey
        };

        // Remove null/undefined params to avoid empty query params
        //Object.keys(params).forEach(key => params[key] == null && delete params[key]);

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the ReverseSearch API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @param {string} Latitude - The latitude of the location (decimal, up to 7 decimal places). Required.
     * @param {string} Longitude - The longitude of the location (decimal, up to 7 decimal places). Required.
     * @param {string} SearchRadius - The maximum search radius in kilometers (1-50). Defaults to '1'.
     * @param {string} Country - The country name or ISO 3166-1 Alpha-2/Alpha-3 code. Optional.
     * @param {string} MaxResults - Maximum number of results (1-10). Defaults to '10'.
     * @param {string} SearchType - Type of search (e.g., 'BestMatch', 'All'). Required.
     * @param {string} LicenseKey - Your license key to use the service. Required.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers. Defaults to true.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service. Defaults to 15.
     * @returns {RSResponse} - An RSResponse object with address or place details or an error.
     */
    invoke(Latitude, Longitude, SearchRadius = '1', Country, MaxResults = '10', SearchType, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            Latitude, Longitude, SearchRadius, Country, MaxResults, SearchType, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { ReverseSearchClient, RSResponse };